File: api/services/rate_limiter.py
import time, os
from fastapi import HTTPException, Request, status
from pydantic import BaseModel
from .redis import get_redis

RATE_LIMIT_REQUESTS = int(os.getenv("RATE_LIMIT_REQUESTS", "100"))
RATE_LIMIT_WINDOW   = int(os.getenv("RATE_LIMIT_WINDOW",   "60"))

class RateLimitInfo(BaseModel):
    limit: int; remaining: int; reset_at: int

async def check_rate_limit(request: Request) -> RateLimitInfo:
    r   = await get_redis()
    ip  = request.client.host if request.client else "unknown"
    key = f"rl:{ip}"
    now = time.time()
    win = now - RATE_LIMIT_WINDOW
    pipe = r.pipeline()
    pipe.zremrangebyscore(key, "-inf", win)
    pipe.zadd(key, {str(now): now})
    pipe.zcard(key)
    pipe.expire(key, RATE_LIMIT_WINDOW)
    results  = await pipe.execute()
    count    = results[2]
    remaining = max(0, RATE_LIMIT_REQUESTS - count)
    reset_at  = int(now) + RATE_LIMIT_WINDOW
    if count > RATE_LIMIT_REQUESTS:
        raise HTTPException(
            status_code=status.HTTP_429_TOO_MANY_REQUESTS,
            detail={"error": "rate_limit_exceeded", "reset_at": reset_at},
            headers={"Retry-After": str(RATE_LIMIT_WINDOW)},
        )
    return RateLimitInfo(limit=RATE_LIMIT_REQUESTS, remaining=remaining, reset_at=reset_at)

File: api/services/rateLimiter.js
import { getRedis } from "./redis.js";

const LIMIT  = Number(process.env.RATE_LIMIT_REQUESTS) || 100;
const WINDOW = Number(process.env.RATE_LIMIT_WINDOW)   || 60;

export async function checkRateLimit(req, res, next) {
  const ip  = req.ip || "unknown";
  const key = `rl:${ip}`;
  const now = Date.now() / 1000;
  const win = now - WINDOW;
  try {
    const r    = getRedis();
    const pipe = r.pipeline();
    pipe.zremrangebyscore(key, "-inf", win);
    pipe.zadd(key, now, String(now));
    pipe.zcard(key);
    pipe.expire(key, WINDOW);
    const results = await pipe.exec();
    const count   = results[2][1];
    const remaining = Math.max(0, LIMIT - count);
    const resetAt   = Math.floor(now) + WINDOW;
    res.set("X-RateLimit-Limit",     String(LIMIT));
    res.set("X-RateLimit-Remaining", String(remaining));
    res.set("X-RateLimit-Reset",     String(resetAt));
    if (count > LIMIT) {
      return res.status(429).json({ error: "rate_limit_exceeded", reset_at: resetAt });
    }
    next();
  } catch { next(); }
}
